pkg/products/threescale/reconciler.go

```diff
@@ -649,138 +648,225 @@ func (r *Reconciler) reconcileComponents(ctx context.Context, serverClient k8scl
-				QueSpec: &threescalev1.ZyncQueSpec{Replicas: &[]int64{0}[0]},
-			},
-		},
-	}
 
 	antiAffinityRequired, err := resources.IsAntiAffinityRequired(ctx, serverClient)
 	if err != nil {
 		r.log.Warning("Failure when deciding if pod anti affinity is required. Defaulted to false: " + err.Error())
 		antiAffinityRequired = false
 	}
 
-	key, err := k8sclient.ObjectKeyFromObject(apim)
-	if err != nil {
-		return integreatlyv1alpha1.PhaseFailed, err
+	apim := &threescalev1.APIManager{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      apiManagerName,
+			Namespace: r.Config.GetNamespace(),
+		},
 	}
 
+	key := k8sclient.ObjectKeyFromObject(apim)
 	err = serverClient.Get(ctx, key, apim)
 	if err != nil && !k8serr.IsNotFound(err) {
 		return integreatlyv1alpha1.PhaseFailed, err
 	}
 
-	status, err := controllerutil.CreateOrUpdate(ctx, serverClient, apim, func() error {
+	replicas := r.Config.GetReplicasConfig(r.installation)
+	systemAppReplicas := replicas["systemApp"]
+	systemSidekiqReplicas := replicas["systemSidekiq"]
+	apicastStageReplicas := replicas["apicastStage"]
+	backendCronReplicas := replicas["backendCron"]
+	zyncReplicas := replicas["zyncApp"]
+	zyncQueReplicas := replicas["zyncQue"]
 
-		apim.Spec.HighAvailability = &threescalev1.HighAvailabilitySpec{Enabled: true}
+	status, err := controllerutil.CreateOrUpdate(ctx, serverClient, apim, func() error {
+		apim.Spec.HighAvailability = &threescalev1.HighAvailabilitySpec{
+			Enabled: true,
+		}
 		apim.Spec.APIManagerCommonSpec.ResourceRequirementsEnabled = &resourceRequirements
-		apim.Spec.APIManagerCommonSpec.WildcardDomain = r.installation.Spec.RoutingSubdomain
-		apim.Spec.System.FileStorageSpec = fss
-		apim.Spec.PodDisruptionBudget = &threescalev1.PodDisruptionBudgetSpec{Enabled: true}
-		apim.Spec.Monitoring = &threescalev1.MonitoringSpec{Enabled: false}
-
-		replicas := r.Config.GetReplicasConfig(r.installation)
-
-		if *apim.Spec.System.AppSpec.Replicas < replicas["systemApp"] {
-			*apim.Spec.System.AppSpec.Replicas = replicas["systemApp"]
+		apim.Spec.WildcardDomain = r.installation.Spec.RoutingSubdomain
+		apim.Spec.ExternalComponents = &threescalev1.ExternalComponentsSpec{
+			System: &threescalev1.ExternalSystemComponents{
+				Redis:    &ExternalComponentsTrue,
+				Database: &ExternalComponentsTrue,
+			},
+			Backend: &threescalev1.ExternalBackendComponents{
+				Redis: &ExternalComponentsTrue,
+			},
 		}
-		if *apim.Spec.System.SidekiqSpec.Replicas < replicas["systemSidekiq"] {
-			*apim.Spec.System.SidekiqSpec.Replicas = replicas["systemSidekiq"]
+
+		if apim.Spec.System != nil {
+			apim.Spec.System.FileStorageSpec = fss
+			apim.Spec.System.AppSpec = &threescalev1.SystemAppSpec{
+				Replicas: apim.Spec.System.AppSpec.Replicas,
+				Affinity: resources.SelectAntiAffinityForCluster(antiAffinityRequired, map[string]string{
+					"threescale_component":         "system",
+					"threescale_component_element": "app",
+				}),
+			}
```

Comment:
When we set these specs could we just modify the field we're changing instead of creating new struct and re-setting the replicas to current value?

```suggestion
			apim.Spec.System.AppSpec.Affinity = resources.SelectAntiAffinityForCluster(
				antiAffinityRequired, map[string]string{
					"threescale_component":         "system",
					"threescale_component_element": "app",
				})
```

Reply:
This is the part of this entire PR that I really don't like but I could not figure out a better solution.
So the problem is that the apim.Spec.System.AppSpec.Affinity will be a nil pointer exception when this PR gets merged. 
Any nested optional field now throws a nil pointer :/

However, I might be missing something so if you have an idea on how to better createOrUpdate the APIManager I'm open to suggestions. 

Best way to verify it locally is by running the unit test: TestReconciler_Reconcile3scale